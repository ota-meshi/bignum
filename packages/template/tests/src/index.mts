import { f } from "../../src/index.mjs";
import * as snap from "@ota-meshi/test-snapshot";

describe("standard tests", () => {
  for (const t of [
    // Doc tests
    // add
    test`0.1 + 0.2`, // 0.3
    // subtract
    test`0.3 - 0.1`, // 0.2
    // multiply
    test`0.07 * 100`, // 7
    // divide
    test`0.6 / 0.2`, // 3
    // modulo
    test`0.6 % 0.2`, // 0
    // pow / nth root
    test`2 ** 3`, // 8
    test`4 ** (1/2)`, // 2
    test`8 ** (1/3)`, // 2
    /* Parentheses */
    test`(0.1 + 0.2) * 10`, // 3
    test`0.1 + 0.2 * 10`, // 2.1
    /* Unary Operators */
    test`${0.3} + -${0.1}`, // 0.2
    /* Logical Operators (Returns 0/1 for value compatibility) */
    test`41 == 41`, // 1
    test`41 == 1`, // 0
    test`41 != 41`, // 0
    test`41 != 1`, // 1
    test`1 <= 2`, // 1
    test`1 <= 1`, // 1
    test`1 <= 0`, // 0
    test`1 < 2`, // 1
    test`1 < 1`, // 0
    test`1 < 1`, // 0
    test`2 >= 1`, // 1
    test`1 >= 1`, // 1
    test`0 >= 1`, // 0
    test`2 > 1`, // 1
    test`1 > 1`, // 0
    test`0 > 1`, // 0

    test`0.1 * 10`, // 1
    test`${0.000002} * ${10}`, // 0.00002
    test`${0.07} * ${100}`, // 7
    test`1 + 2`,
    test`${0.1} + 0.1 * 2`,
    test`1+2`,
    test`1.1 + 2.1`,
    test`1+-2`,
    test`1-+2`,
    test`1.1+-2.2`,
    test`1.2+0.8`,
    test`1 + ${3}`,
    test`1 + ${2} * 3`,
    test`1.2 + ${2.3} * 3.4`,
    test`(1 + ${2}) * 3`,
    test`1 * -(23 % 4)`,
    test`1.5 * -(23 % 4)`,
    test`3*3*PI`,
    test`3.2*3.2*PI`,
    test`trunc(300*1.08)`,
    test`trunc(320*1.08)`,
    test`1 + 2 * 3`,
    test`3 - 3 / 3`,
    test`3.2 - 3.2 / 3.2`,
    test`(1 + 2) * 3`,
    test`2 * 2 ** 3`,
    test`2.1 * 2.3 ** 3`,
    test`2.1 * 2.3 ** (1.5+1.5)`,
    // test`2.1 * 2.3 ** 3.4`, // Error
    test`4 / 2 ** 2`,
    test`4.1 / 2.2 ** 2`,
    test`(2 * 2) ** 3`,
    test`abs(-10)`,
    test`(2 * 2) / 10 ** 20`,
    test`(2 * 2) / 10 ** 21`, // Overflow the default maximum number of decimal places. (20).
    test`2.3 / 1.1`,
    test`2.3 % 1.1`,
    test`22000000 / 1.1`,
    ...[
      "1.49",
      "1.5",
      "1.51",
      "1.501",
      "-1.49",
      "-1.5",
      "-1.51",
      "-1.501",
    ].flatMap((v) => [
      test`trunc(${v})`,
      test`round(${v})`,
      test`floor(${v})`,
      test`ceil(${v})`,
      test`abs(${v})`,
    ]),
    test`1e42+1`,
    test`1e-42+1`,
    test`1e+42+1`,
    test`1.23e42+1`,
    test`1.23e-42+1`,
    test`1.23e+42+1`,
    test`-1.23e42+1`,
    test`-1.23e-42+1`,
    test`-1.23e+42+1`,
    ...["0", "1", "1.49", "1.5", "1.51", "1.501", "2"].flatMap((v) => [
      test`sqrt(${v})`,
    ]),
  ]) {
    it(t.name, () => {
      snap.expect(f(...t.param)).toMatchSnapshot();
    });
  }
});
describe("Big Number tests", () => {
  for (const t of [
    test`0.1 + 0.1 + 0.1`,
    test`${0.1} + ${0.1} + ${0.1}`,
    test`0.2 + 0.1`,
    test`${0.2} + ${0.1}`,
    test`0.3 - 0.1`,
    test`${0.3} - ${0.1}`,
  ]) {
    it(t.name, () => {
      snap.expect(f(...t.param)).toMatchSnapshot();
    });
  }
});
describe("Compare tests", () => {
  for (const t of [
    test`0.1 + 0.1 + 0.1 <= 0.3`,
    test`${0.1} + ${0.1} + ${0.1} == ${0.3}`,
    test`0.2 + 0.1 != 0.3`,
    test`${0.2} + ${0.1} != ${0.3}`,
    test`0.3 - 0.1 >= 0.2`,
    test`${0.3} - ${0.1} >= ${0.2}`,
    test`${0.3} - ${0.1} < ${0.2}`,
    test`${0.3} - ${0.1} > ${0.2}`,
    ...[
      [0.1, 0.1],
      [0.1, 0.2],
      [0.2, 0.1],
    ].flatMap(([a, b]) => [
      test`${a} < ${b}`,
      test`${a} <= ${b}`,
      test`${a} > ${b}`,
      test`${a} >= ${b}`,
      test`${a} == ${b}`,
      test`${a} != ${b}`,
    ]),
  ]) {
    it(t.name, () => {
      snap.expect(f(...t.param)).toMatchSnapshot();
    });
  }
});

/**
 *
 */
function test(
  template: TemplateStringsArray,
  ...substitutions: (number | bigint | string)[]
) {
  let name = "`";
  for (let index = 0; index < template.length; index++) {
    name += template[index];
    if (index < substitutions.length) {
      name += `\${${substitutions[index]}}`;
    }
  }
  name += "`";
  return {
    name,
    param: [template, ...substitutions] as const,
  };
}
