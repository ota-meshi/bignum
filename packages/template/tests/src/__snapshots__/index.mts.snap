// test-snapshot v1

exports[`Big Number tests \`\${0.1} + \${0.1} + \${0.1}\` 1`] = `0.3`;

exports[`Big Number tests \`\${0.2} + \${0.1}\` 1`] = `0.3`;

exports[`Big Number tests \`\${0.3} - \${0.1}\` 1`] = `0.2`;

exports[`Big Number tests \`0.1 + 0.1 + 0.1\` 1`] = `0.3`;

exports[`Big Number tests \`0.2 + 0.1\` 1`] = `0.3`;

exports[`Big Number tests \`0.3 - 0.1\` 1`] = `0.2`;

exports[`Compare tests \`\${0.1} != \${0.1}\` 1`] = `0`;

exports[`Compare tests \`\${0.1} != \${0.2}\` 1`] = `1`;

exports[`Compare tests \`\${0.1} + \${0.1} + \${0.1} == \${0.3}\` 1`] = `1`;

exports[`Compare tests \`\${0.1} < \${0.1}\` 1`] = `0`;

exports[`Compare tests \`\${0.1} < \${0.2}\` 1`] = `1`;

exports[`Compare tests \`\${0.1} <= \${0.1}\` 1`] = `1`;

exports[`Compare tests \`\${0.1} <= \${0.2}\` 1`] = `1`;

exports[`Compare tests \`\${0.1} == \${0.1}\` 1`] = `1`;

exports[`Compare tests \`\${0.1} == \${0.2}\` 1`] = `0`;

exports[`Compare tests \`\${0.1} > \${0.1}\` 1`] = `0`;

exports[`Compare tests \`\${0.1} > \${0.2}\` 1`] = `0`;

exports[`Compare tests \`\${0.1} >= \${0.1}\` 1`] = `1`;

exports[`Compare tests \`\${0.1} >= \${0.2}\` 1`] = `0`;

exports[`Compare tests \`\${0.2} != \${0.1}\` 1`] = `1`;

exports[`Compare tests \`\${0.2} + \${0.1} != \${0.3}\` 1`] = `0`;

exports[`Compare tests \`\${0.2} < \${0.1}\` 1`] = `0`;

exports[`Compare tests \`\${0.2} <= \${0.1}\` 1`] = `0`;

exports[`Compare tests \`\${0.2} == \${0.1}\` 1`] = `0`;

exports[`Compare tests \`\${0.2} > \${0.1}\` 1`] = `1`;

exports[`Compare tests \`\${0.2} >= \${0.1}\` 1`] = `1`;

exports[`Compare tests \`\${0.3} - \${0.1} < \${0.2}\` 1`] = `0`;

exports[`Compare tests \`\${0.3} - \${0.1} > \${0.2}\` 1`] = `0`;

exports[`Compare tests \`\${0.3} - \${0.1} >= \${0.2}\` 1`] = `1`;

exports[`Compare tests \`0.1 + 0.1 + 0.1 <= 0.3\` 1`] = `1`;

exports[`Compare tests \`0.2 + 0.1 != 0.3\` 1`] = `0`;

exports[`Compare tests \`0.3 - 0.1 >= 0.2\` 1`] = `1`;

exports[`standard tests \`\${0.000002} * \${10}\` 1`] = `0.00002`;

exports[`standard tests \`\${0.07} * \${100}\` 1`] = `7`;

exports[`standard tests \`\${0.1} + 0.1 * 2\` 1`] = `0.3`;

exports[`standard tests \`\${0.3} + -\${0.1}\` 1`] = `0.2`;

exports[`standard tests \`(0.1 + 0.2) * 10\` 1`] = `3`;

exports[`standard tests \`(1 + \${2}) * 3\` 1`] = `9`;

exports[`standard tests \`(1 + 2) * 3\` 1`] = `9`;

exports[`standard tests \`(2 * 2) ** 3\` 1`] = `64`;

exports[`standard tests \`(2 * 2) / 10 ** 20\` 1`] = `"0.00000000000000000004"`;

exports[`standard tests \`(2 * 2) / 10 ** 21\` 1`] = `"0.000000000000000000004"`;

exports[`standard tests \`-1.23e+42+1\` 1`] = `"-1229999999999999999999999999999999999999999"`;

exports[`standard tests \`-1.23e-42+1\` 1`] = `"0.99999999999999999999999999999999999999999877"`;

exports[`standard tests \`-1.23e42+1\` 1`] = `"-1229999999999999999999999999999999999999999"`;

exports[`standard tests \`0 > 1\` 1`] = `0`;

exports[`standard tests \`0 >= 1\` 1`] = `0`;

exports[`standard tests \`0.07 * 100\` 1`] = `7`;

exports[`standard tests \`0.1 * 10\` 1`] = `1`;

exports[`standard tests \`0.1 + 0.2 * 10\` 1`] = `2.1`;

exports[`standard tests \`0.1 + 0.2\` 1`] = `0.3`;

exports[`standard tests \`0.3 - 0.1\` 1`] = `0.2`;

exports[`standard tests \`0.6 % 0.2\` 1`] = `0`;

exports[`standard tests \`0.6 / 0.2\` 1`] = `3`;

exports[`standard tests \`1 * -(23 % 4)\` 1`] = `-3`;

exports[`standard tests \`1 + \${2} * 3\` 1`] = `7`;

exports[`standard tests \`1 + \${3}\` 1`] = `4`;

exports[`standard tests \`1 + 2 * 3\` 1`] = `7`;

exports[`standard tests \`1 + 2\` 1`] = `3`;

exports[`standard tests \`1 < 1\` 1`] = `0`;

exports[`standard tests \`1 < 1\` 2`] = `0`;

exports[`standard tests \`1 < 2\` 1`] = `1`;

exports[`standard tests \`1 <= 0\` 1`] = `0`;

exports[`standard tests \`1 <= 1\` 1`] = `1`;

exports[`standard tests \`1 <= 2\` 1`] = `1`;

exports[`standard tests \`1 > 1\` 1`] = `0`;

exports[`standard tests \`1 >= 1\` 1`] = `1`;

exports[`standard tests \`1+-2\` 1`] = `-1`;

exports[`standard tests \`1+2\` 1`] = `3`;

exports[`standard tests \`1.1 + 2.1\` 1`] = `3.2`;

exports[`standard tests \`1.1+-2.2\` 1`] = `-1.1`;

exports[`standard tests \`1.2 + \${2.3} * 3.4\` 1`] = `9.02`;

exports[`standard tests \`1.2+0.8\` 1`] = `2`;

exports[`standard tests \`1.5 * -(23 % 4)\` 1`] = `-4.5`;

exports[`standard tests \`1.23e+42+1\` 1`] = `"1230000000000000000000000000000000000000001"`;

exports[`standard tests \`1.23e-42+1\` 1`] = `"1.00000000000000000000000000000000000000000123"`;

exports[`standard tests \`1.23e42+1\` 1`] = `"1230000000000000000000000000000000000000001"`;

exports[`standard tests \`1-+2\` 1`] = `-1`;

exports[`standard tests \`1e+42+1\` 1`] = `"1000000000000000000000000000000000000000001"`;

exports[`standard tests \`1e-42+1\` 1`] = `"1.000000000000000000000000000000000000000001"`;

exports[`standard tests \`1e42+1\` 1`] = `"1000000000000000000000000000000000000000001"`;

exports[`standard tests \`2 * 2 ** 3\` 1`] = `16`;

exports[`standard tests \`2 ** 3\` 1`] = `8`;

exports[`standard tests \`2 > 1\` 1`] = `1`;

exports[`standard tests \`2 >= 1\` 1`] = `1`;

exports[`standard tests \`2.1 * 2.3 ** (1.5+1.5)\` 1`] = `25.5507`;

exports[`standard tests \`2.1 * 2.3 ** 3\` 1`] = `25.5507`;

exports[`standard tests \`2.3 % 1.1\` 1`] = `0.1`;

exports[`standard tests \`2.3 / 1.1\` 1`] = `"2.090909090909090909"`;

exports[`standard tests \`3 - 3 / 3\` 1`] = `2`;

exports[`standard tests \`3*3*PI\` 1`] = `"28.274333882308137"`;

exports[`standard tests \`3.2 - 3.2 / 3.2\` 1`] = `2.2`;

exports[`standard tests \`3.2*3.2*PI\` 1`] = `"32.16990877275948032"`;

exports[`standard tests \`4 ** (1/2)\` 1`] = `2`;

exports[`standard tests \`4 / 2 ** 2\` 1`] = `1`;

exports[`standard tests \`4.1 / 2.2 ** 2\` 1`] = `"0.84710743801652892561"`;

exports[`standard tests \`8 ** (1/3)\` 1`] = `2`;

exports[`standard tests \`41 != 1\` 1`] = `1`;

exports[`standard tests \`41 != 41\` 1`] = `0`;

exports[`standard tests \`41 == 1\` 1`] = `0`;

exports[`standard tests \`41 == 41\` 1`] = `1`;

exports[`standard tests \`22000000 / 1.1\` 1`] = `20000000`;

exports[`standard tests \`abs(\${-1.5})\` 1`] = `1.5`;

exports[`standard tests \`abs(\${-1.49})\` 1`] = `1.49`;

exports[`standard tests \`abs(\${-1.51})\` 1`] = `1.51`;

exports[`standard tests \`abs(\${-1.501})\` 1`] = `1.501`;

exports[`standard tests \`abs(\${1.5})\` 1`] = `1.5`;

exports[`standard tests \`abs(\${1.49})\` 1`] = `1.49`;

exports[`standard tests \`abs(\${1.51})\` 1`] = `1.51`;

exports[`standard tests \`abs(\${1.501})\` 1`] = `1.501`;

exports[`standard tests \`abs(-10)\` 1`] = `10`;

exports[`standard tests \`ceil(\${-1.5})\` 1`] = `-1`;

exports[`standard tests \`ceil(\${-1.49})\` 1`] = `-1`;

exports[`standard tests \`ceil(\${-1.51})\` 1`] = `-1`;

exports[`standard tests \`ceil(\${-1.501})\` 1`] = `-1`;

exports[`standard tests \`ceil(\${1.5})\` 1`] = `2`;

exports[`standard tests \`ceil(\${1.49})\` 1`] = `2`;

exports[`standard tests \`ceil(\${1.51})\` 1`] = `2`;

exports[`standard tests \`ceil(\${1.501})\` 1`] = `2`;

exports[`standard tests \`floor(\${-1.5})\` 1`] = `-2`;

exports[`standard tests \`floor(\${-1.49})\` 1`] = `-2`;

exports[`standard tests \`floor(\${-1.51})\` 1`] = `-2`;

exports[`standard tests \`floor(\${-1.501})\` 1`] = `-2`;

exports[`standard tests \`floor(\${1.5})\` 1`] = `1`;

exports[`standard tests \`floor(\${1.49})\` 1`] = `1`;

exports[`standard tests \`floor(\${1.51})\` 1`] = `1`;

exports[`standard tests \`floor(\${1.501})\` 1`] = `1`;

exports[`standard tests \`round(\${-1.5})\` 1`] = `-1`;

exports[`standard tests \`round(\${-1.49})\` 1`] = `-1`;

exports[`standard tests \`round(\${-1.51})\` 1`] = `-2`;

exports[`standard tests \`round(\${-1.501})\` 1`] = `-2`;

exports[`standard tests \`round(\${1.5})\` 1`] = `2`;

exports[`standard tests \`round(\${1.49})\` 1`] = `1`;

exports[`standard tests \`round(\${1.51})\` 1`] = `2`;

exports[`standard tests \`round(\${1.501})\` 1`] = `2`;

exports[`standard tests \`sqrt(\${0})\` 1`] = `0`;

exports[`standard tests \`sqrt(\${1.5})\` 1`] = `"1.224744871391589049"`;

exports[`standard tests \`sqrt(\${1.49})\` 1`] = `"1.2206555615733702951"`;

exports[`standard tests \`sqrt(\${1.51})\` 1`] = `"1.2288205727444507591"`;

exports[`standard tests \`sqrt(\${1.501})\` 1`] = `"1.2251530516633421829"`;

exports[`standard tests \`sqrt(\${1})\` 1`] = `1`;

exports[`standard tests \`sqrt(\${2})\` 1`] = `"1.4142135623730950488"`;

exports[`standard tests \`trunc(\${-1.5})\` 1`] = `-1`;

exports[`standard tests \`trunc(\${-1.49})\` 1`] = `-1`;

exports[`standard tests \`trunc(\${-1.51})\` 1`] = `-1`;

exports[`standard tests \`trunc(\${-1.501})\` 1`] = `-1`;

exports[`standard tests \`trunc(\${1.5})\` 1`] = `1`;

exports[`standard tests \`trunc(\${1.49})\` 1`] = `1`;

exports[`standard tests \`trunc(\${1.51})\` 1`] = `1`;

exports[`standard tests \`trunc(\${1.501})\` 1`] = `1`;

exports[`standard tests \`trunc(300*1.08)\` 1`] = `324`;

exports[`standard tests \`trunc(320*1.08)\` 1`] = `345`;
